samza-core/src/main/java/org/apache/samza/operators/impl/OutputOperatorImpl.java

```diff
@@ -69,4 +69,14 @@ protected void handleClose() {
   protected OperatorSpec<M, Void> getOperatorSpec() {
     return outputOpSpec;
   }
+
+  @Override
+  protected long handleWatermark(Watermark inputWatermark,
+      MessageCollector collector,
+      TaskCoordinator coordinator) {
+    if (outputOpSpec.getOpCode() == OperatorSpec.OpCode.PARTITION_BY) {
+      inputWatermark.propagate(outputStream.getStreamSpec().toSystemStream());
```

Comment:
Ideally when we propagate we should not need to be aware of who we are propagating to. This should be internally managed (and again, probably not on the watermark object).